addons/api/addon/handlers/indexed-db-handler.js

```diff
@@ -0,0 +1,119 @@
+import { modelIndexes } from 'api/services/indexed-db';
+import { inject as service } from '@ember/service';
+import { getOwner, setOwner } from '@ember/application';
+import { pluralize } from 'ember-inflector';
+
+/**
+ * Handler to sit in front of the API request layer
+ * so we can handle any caching with indexedDb first
+ */
+export default class IndexedDbHandler {
+  @service indexedDb;
+
+  constructor(context) {
+    setOwner(this, getOwner(context));
+  }
+
+  async request(context, next) {
+    switch (context.request.op) {
+      case 'query': {
+        const { store, data } = context.request;
+        const { type, query, options: { pushToStore = true } = {} } = data;
+        const supportedModels = Object.keys(modelIndexes);
+        const { db: indexedDb } = this.indexedDb ?? {};
+        const pluralizedType = pluralize(type);
+
+        // Go through normal flow if we don't yet support the model
+        // or if we don't have an indexedDb instance
+        if (!supportedModels.includes(pluralizedType) || !indexedDb) {
+          return next(context.request);
+        }
+
+        const listToken = await indexedDb.tokens.get(pluralizedType);
+        const adapter = store.adapterFor(type);
+        const schema = store.modelFor(type);
+
+        let payload;
+        try {
+          payload = await adapter.query(store, schema, {
+            ...query,
+            scope_id: 'global',
+            recursive: true,
+            list_token: listToken?.token,
+          });
+        } catch (e) {
+          // If we get any error that's not a rate limiting error,
+          // assume the token is not good anymore. We'll delete the token and
+          // clear the DB and try again without a token.
+          // TODO: Probably should check for 400 and bad token error specifically instead?
+          if (e?.errors[0].status !== 429) {
+            await indexedDb[pluralizedType].clear();
+            await indexedDb.tokens.delete(pluralizedType);
+
+            payload = await adapter.query(store, schema, {
+              ...query,
+              scope_id: 'global',
+              recursive: true,
+              list_token: null,
+            });
+          }
+        }
```

Comment:
im not completely understanding what should happen in the case of there being a rate limit error. So right now if there's a rate limit error, we just continue but further down in line 85 we do a bulkPut operation. Im thinking that maybe that should be skipped in case of rate limit error, or what is the reason for storing the payload?

Reply:
Yeah it's not great because if we do just get an error, there's not much we _can_ do depending on the error. This check really should only be for a bad token error which should get updated later but if we do get any other error including rate limiting, we probably want to still get the partial results before being rate limited and still update the DB, but otherwise we most likely get nothing and need to bail.

You're probably right and if we do get an error we should just return immediately here, most likely I should just re-throw any thing that's not a bad token error, thoughts?